### ISSUE TITLE
Function pointer types produce invalid Solidity output

### ISSUE THREAD

**Issue by gnidan** [gnidan/abi-to-sol#76]
Originally raised by @esaulpaugh in #67.

For type `"function"` parameters, abi-to-sol only outputs `function <identifier>` rather than proper Solidity syntax which includes arguments and return type.

e.g., the following Solidity:
<details>
<summary>Map.sol</summary>

```solidity
// SPDX-License-Identifier: UNLICENSED

pragma solidity ^0.8.11;

contract Map {
    function map(
        function (uint256) external pure returns (uint256) f,
        uint256[] calldata list
    )
        public
        pure
        returns (uint256[] memory)
    {
        uint256[] memory mapped = new uint256[](list.length);

        for (uint256 i = 0; i < list.length; i++) {
            mapped[i] = f(list[i]);
        }

        return mapped;
    }
}
```

</details>

This contract defines `function map(function (uint256) external pure returns (uint256) f, uint256[])` to accept an external function pointer as argument. As @esaulpaugh has also stated in the other issue, the output ABI JSON for this contract defines `type` and `internalType`, where `type` specifies merely `function` and no other information, and `internalType` _does_ provide enough information but is not spec-compliant.

See full ABI JSON:
<details>
<summary>Corresponding ABI JSON for contract `Map`</summary>

```json
[
  {
    "inputs": [
      {
        "internalType": "function (uint256) pure external returns (uint256)",
        "name": "f",
        "type": "function"
      },
      {
        "internalType": "uint256[]",
        "name": "list",
        "type": "uint256[]"
      }
    ],
    "name": "map",
    "outputs": [
      {
        "internalType": "uint256[]",
        "name": "",
        "type": "uint256[]"
      }
    ],
    "stateMutability": "pure",
    "type": "function"
  }
]
```

</details>

And **see what abi-to-sol produces**:
<details>
<summary>abi-to-sol output for this ABI JSON (hand-prettified)</summary>

```solidity
// SPDX-License-Identifier: UNLICENSED
// !! THIS FILE WAS AUTOGENERATED BY abi-to-sol v0.6.2. SEE SOURCE BELOW. !!
pragma solidity >=0.7.0 <0.9.0;

interface MyInterface {
    function map(
        function f,
        uint256[] memory list
    ) external pure returns (uint256[] memory);
}

// THIS FILE WAS AUTOGENERATED FROM THE FOLLOWING ABI JSON:
/*
[{"inputs":[{"internalType":"function (uint256) pure external returns (uint256)","name":"f","type":"function"},{"internalType":"uint256[]","name":"list","type":"uint256[]"}],"name":"map","outputs":[{"internalType":"uint256[]","name":"","type":"uint256[]"}],"stateMutability":"pure","type":"function"}]
*/
```

</details>

Since abi-to-sol already uses `internalType` extensively when it is available, the trick here, I suppose, will be to identify appropriate fallback behavior when `internalType` is not available... should abi-to-sol just fail in this situation? Or maybe it's fine? I'll have to do some testing.

Anyway, thanks @esaulpaugh for bringing this to my attention!

**Comment by gnidan** [comment]
Consideration for what to do about input ABI JSONs where `internalType` isn't provided:

If abi-to-sol outputs just `function() f` for the parameter, then in theory one could use `abi.encodePacked()` alongside Solidity's `f.address` and `f.selector` syntax (defined in their [Function types documentation](https://docs.soliditylang.org/en/v0.8.15/types.html#function-types)) to interact with the function correctly. This would require actually knowing the parameter and return types ahead of time, and there'd be no type safety, but it's probably still better than throwing an exception. Maybe abi-to-sol should output a warning comment inline in this situation?

This is what I'll go with when I implement the fix for this.

**Comment by gnidan** [comment]
This is now fixed and released in [abi-to-sol v0.6.3](https://github.com/gnidan/abi-to-sol/releases/tag/v0.6.3) (and deployed to the gh-page). Thanks again!
